ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/SystemColumns.java

```diff
@@ -15,39 +15,78 @@
 
 package io.confluent.ksql.schema.ksql;
 
+import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
 import io.confluent.ksql.name.ColumnName;
 import io.confluent.ksql.schema.ksql.types.SqlType;
 import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.util.KsqlException;
+import java.util.HashSet;
+import java.util.Map;
 import java.util.Set;
 
 public final class SystemColumns {
 
   public static final ColumnName ROWKEY_NAME = ColumnName.of("ROWKEY");
-
   public static final ColumnName ROWTIME_NAME = ColumnName.of("ROWTIME");
   public static final SqlType ROWTIME_TYPE = SqlTypes.BIGINT;
 
+  public static final ColumnName ROWOFFSET_NAME = ColumnName.of("ROWOFFSET");
+  public static final SqlType ROWOFFSET_TYPE = SqlTypes.BIGINT;
+
+  public static final ColumnName ROWPARTITION_NAME = ColumnName.of("ROWPARTITION");
+  public static final SqlType ROWPARTITION_TYPE = SqlTypes.INTEGER;
+
   public static final ColumnName WINDOWSTART_NAME = ColumnName.of("WINDOWSTART");
   public static final ColumnName WINDOWEND_NAME = ColumnName.of("WINDOWEND");
   public static final SqlType WINDOWBOUND_TYPE = SqlTypes.BIGINT;
 
-  public static final int LEGACY_PSEUDOCOLUMN_VERSION_NUMBER = 0;
-  public static final int CURRENT_PSEUDOCOLUMN_VERSION_NUMBER = 5;
 
-  private static final Set<ColumnName> PSEUDO_COLUMN_NAMES = ImmutableSet.of(
-      ROWTIME_NAME
-  );
+  public static final int ROWTIME_PSEUDOCOLUMN_VERSION = 0;
+  public static final int ROWPARTITION_ROWOFFSET_PSEUDOCOLUMN_VERSION = 1;
+
+  public static final int LEGACY_PSEUDOCOLUMN_VERSION_NUMBER = ROWTIME_PSEUDOCOLUMN_VERSION;
+  public static final int CURRENT_PSEUDOCOLUMN_VERSION_NUMBER =
+      ROWPARTITION_ROWOFFSET_PSEUDOCOLUMN_VERSION;
+
+  public static Set<ColumnName> getPseudoColumnsFromVersion(final int pseudoColumnVersion) {
+    if (!VERSIONED_PSEUDO_COLUMNS.containsKey(pseudoColumnVersion)) {
+      throw new KsqlException("Provided pseudoColumnVersion has no corresponding columns defined");
+    }
+    return VERSIONED_PSEUDO_COLUMNS.get(pseudoColumnVersion);
+  }
+
+  private static final Set<ColumnName> VERSION_ZERO_NAMES;
```

Comment:
We've added a lot of pseudocolumn-specific code at this point this is making this file (SystemColumns.java) fairly difficult to read. I think we should create a new util file specifically for pseudocolumns, e.g., `PseudoColumns.java`.

Reply:
Sounds good, the only concern I have is that this will add interdependencies for functions like `isSystemColumn()` and `systemColumnNames()`, as Pseudo Columns are a subset of System Columns. I could make a set of package-private functions in `PseudoColumns.java` that allows SystemColumns to access this internal data as necessary. A problem with this though is I think that this internal will pretty much always be necessary, and the resulting code will be more complex  as `SystemColumns` will end up being more complex than it was before and we'll have another class, `PseudoColumns`, which is almost as complex as `SystemColumns` is now.

I'll try to prototype this change before our sync to see if my concerns are well-founded or not.

Reply:
Another option is to group all the pseudo-column specific code into a static class within SystemColumns.java, rather than creating a separate file. I think either is fine. The interdependencies in the case of having a separate file would be package-private rather than publicly accessible.